### PR TITLE
fix aliasing bug in pixel shuffle/unshuffle

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/82235

cc @albanD - `at::pixel_shuffle` and `at::pixel_unshuffle` advertise as being non-aliasing, but they have a C++ decomposition that internally uses reshape(), which means that it might return an alias.

I happened to notice this because a bunch of tests in `test/test_ops.py` failed when I ran locally with a `DEBUG=1` build.

(P.S.: when are we finally gonna get a debug build test in CI? 😃)

I fixed by adding an extra clone, which... is going to be an unnecessary perf hit in the case where the `reshape()` already properly cloned the input. My hope is that this is fine, because this only impacts the composite kernel- we already have a "fast" CPU kernel that does the right thing. Is `pixel_shuffle/unshuffle` commonly used with cuda? Maybe we should just add a fast cuda kernel for it if that's the case.

Alternatively, it seems like it would be nice if `reshape()` accepted an optional argument to unconditionally return a copy. That seems like a rabbit hole that isn't worth going down for now though - I remember a discussion a while ago about making `reshape()` copy-on-write

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86611
* #86610
* #86609
* #86228
* __->__ #86608
* #86607
* #86285

### CODE REVIEW

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 1]
```diff
@@ -129,7 +129,8 @@ Tensor math_pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
   std::vector<int64_t> final_shape(self.sizes().begin(), self_sizes_batch_end);
   final_shape.insert(final_shape.end(), {oc, oh, ow});
 
-  return input_permuted.reshape(final_shape);
+  // pixel_shuffle expects to *never* return an alias of the input.
+  return input_permuted.reshape(final_shape).clone();
```
Comment (albanD) [inline aten/src/ATen/native/PixelShuffle.cpp:133]: Just do: `input_permuted.clone().view(final_shape)` to avoid the extra copy ;)

Comment (bdhirsh) [reply aten/src/ATen/native/PixelShuffle.cpp:133]: oh you're totally right haha

**Review by albanD — APPROVED** [review]
SGTM

**Review by eellison — COMMENTED** [review]
Cool! Would you remove the skips here ? https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L1778

**Comment by vadimkantorov** [comment]
about extra copy:
there's a common theme that sometimes non-strict "aliasing is possible" (in reshape-style) semantics is preferable to elide copies in cases where it's known that no inplace will happen. e.g. one in-the-wild examples is torch.cat wrapper in detectron2:
https://github.com/facebookresearch/detectron2/blob/48b598b4f61fbb24182a69b521b2a0ba3252b842/detectron2/layers/wrappers.py#L39

that shortcuts if the number of passed tensors is 1

(similar cases may be multiplying by 1 / adding 0, but they might be less realistic...)

so not sure how to control for this - if deemed important aspect. maybe allow copy=True/copy=None arguments to such ops with copy=True by default

**Comment by vadimkantorov** [comment]
Somehow, one way could be to allow express "no inplace ops are allowed on this tensor or its views" or at least "assume no inplace ops will happen on this tensor or its views", then the executor engine can optimize out the unneeded copies easily @albanD
